packages/core/src/controllers/events.ts

```diff
@@ -0,0 +1,200 @@
+export class EventClient extends IEventClient {
+  public readonly context = EVENTS_STORAGE_CONTEXT;
+  private readonly storagePrefix = CORE_STORAGE_PREFIX;
+  private readonly storageVersion = EVENTS_STORAGE_VERSION;
+
+  private events = new Map<string, EventClientTypes.Event>();
+  private toPersist = false;
+  constructor(public core: ICore, public logger: Logger, telemetryEnabled = true) {
+    super(core, logger, telemetryEnabled);
+    this.logger = generateChildLogger(logger, this.context);
+    if (telemetryEnabled) {
+      this.restore().then(async () => {
+        await this.submit();
+        this.setEventListeners();
+      });
+    } else {
+      // overwrite any persisted events with an empty array
+      this.persist();
+    }
+  }
+
+  get storageKey() {
+    return (
+      this.storagePrefix + this.storageVersion + this.core.customStoragePrefix + "//" + this.context
+    );
+  }
+
+  public createEvent: IEventClient["createEvent"] = (params) => {
+    const {
+      event = "ERROR",
+      type = "",
+      properties: { topic, trace },
+    } = params;
+    const eventId = uuidv4();
+    const bundleId = this.core.projectId || "";
+    const timestamp = Date.now();
+    const props = {
+      event,
+      type,
+      properties: {
+        topic,
+        trace,
+      },
+    };
+    const eventObj = {
+      eventId,
+      bundleId,
+      timestamp,
+      props,
+      ...this.setMethods(eventId),
+    };
+
+    if (this.telemetryEnabled) {
+      this.events.set(eventId, eventObj);
+      this.toPersist = true;
+    }
+
+    return eventObj;
+  };
+
+  public getEvent: IEventClient["getEvent"] = (params) => {
+    const { eventId, topic } = params;
+    if (eventId) {
+      return this.events.get(eventId);
+    }
+    const event = Array.from(this.events.values()).find(
+      (event) => event.props.properties.topic === topic,
+    );
+
+    if (!event) return;
+
+    return {
+      ...event,
+      ...this.setMethods(event.eventId),
+    };
+  };
+
+  public deleteEvent: IEventClient["deleteEvent"] = (params) => {
+    const { eventId } = params;
+    this.events.delete(eventId);
+    this.toPersist = true;
+  };
+
+  private setEventListeners = () => {
+    this.core.heartbeat.on(HEARTBEAT_EVENTS.pulse, async () => {
+      if (this.toPersist) await this.persist();
+      // cleanup events older than EVENTS_STORAGE_CLEANUP_INTERVAL
+      this.events.forEach((event) => {
+        if (
+          fromMiliseconds(Date.now()) - fromMiliseconds(event.timestamp) >
+          EVENTS_STORAGE_CLEANUP_INTERVAL
+        ) {
+          this.events.delete(event.eventId);
+          this.toPersist = true;
+        }
+      });
+    });
+  };
+
+  private setMethods = (eventId: string) => {
+    return {
+      addTrace: (trace: string) => this.addTrace(eventId, trace),
+      setError: (errorType: string) => this.setError(eventId, errorType),
+    };
+  };
+
+  private addTrace = async (eventId: string, trace: string) => {
```

Comment:
Why are `addTrace` and `setError` marked as `async`? This makes me think they might be long-running. I think we can remove this to guarantee they don't block the flows.

Reply:
I don't understand why those are returning promises either.